sdk/internal/testframework/recording.go

```diff
@@ -0,0 +1,373 @@
+	alphanumericLowercaseBytes  = "abcdefghijklmnopqrstuvwxyz1234567890"
+	randomSeedVariableName      = "randomSeed"
+	ModeEnvironmentVariableName = "AZURE_TEST_MODE"
+)
+
+// Inspired by https://stackoverflow.com/questions/22892120/how-to-generate-a-random-string-of-a-fixed-length-in-go
+const (
+	letterIdxBits = 6                    // 6 bits to represent a letter index
+	letterIdxMask = 1<<letterIdxBits - 1 // All 1-bits, as many as letterIdxBits
+	letterIdxMax  = 63 / letterIdxBits   // # of letter indices fitting in 63 bits
+)
+
+type RecordMode string
+
+const (
+	Record   RecordMode = "record"
+	Playback RecordMode = "playback"
+	Live     RecordMode = "live"
+)
+
+type VariableType string
+
+const (
+	Default             VariableType = "default"
+	Secret_String       VariableType = "secret_string"
+	Secret_Base64String VariableType = "secret_base64String"
+)
+
+func NewRecording(c TestContext, mode RecordMode) (*Recording, error) {
+	// create recorder based on the test name, recordMode, variables, and sanitizers
+	recPath, varPath := getFilePaths(c.Name())
+	rec, err := recorder.NewAsMode(recPath, modeMap[mode], nil)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// If the mode is set in the environment, let that override the requested mode
+	// This is to enable support for nightly live test pipelines
+	envMode := GetOptionalEnv(ModeEnvironmentVariableName, string(mode))
+	mode = RecordMode(*envMode)
+
+	// initialize the Recording
+	recording := &Recording{
+		SessionName:              recPath,
+		RecordingFile:            recPath + ".yaml",
+		VariablesFile:            varPath,
+		Mode:                     mode,
+		variables:                make(map[string]*string),
+		previousSessionVariables: make(map[string]*string),
+		recorder:                 rec,
+		c:                        c,
+	}
+
+	// Try loading the recording if it already exists to hydrate the variables
+	err = recording.initVariables()
+	if err != nil {
+		return nil, err
+	}
+
+	// set the recorder Matcher
+	rec.SetMatcher(recording.matchRequest)
+
+	// wire up the sanitizer
+	DefaultSanitizer(rec)
+
+	return recording, err
+}
+
+// Gets a recorded variable. If the variable is not found we return an error
+// variableType is optional and defaults to Default
+func (r *Recording) GetRecordedVariable(name string, variableType ...VariableType) (*string, error) {
+	var err error
+	result, ok := r.previousSessionVariables[name]
+	if !ok || r.Mode == Live {
+
+		result, err = GetRequiredEnv(name)
+		if err != nil {
+			r.c.Fail(err.Error())
+			return nil, err
+		} else {
+			r.variables[name] = applyVariableOptions(result, variableType...)
+		}
+	}
+	return result, err
+}
+
+// Gets a recorded variable with a fallback default value
+// variableType is optional and defaults to Default
+func (r *Recording) GetOptionalRecordedVariable(name string, defaultValue string, variableType ...VariableType) string {
```

Comment:
I think I was just trying to be clever here. Making it a pointer would also work, but I liked the idea of callers not having to specify it unless it was a secret value. This is probably not very idiomatic though.

Reply:
The variadic config is idiomatic, I just dont think it is in this case. 🤔 This popped out to me, because when I saw the variadic config I thought multiple transforms would be applied to the variable type. If you think it'll always remain a none or 1 choice then I'd say you could just make it a pointer BUT if at any point there could be more then this would be fine. We could discuss this in the meeting next week.

Reply:
As I've thought about it more, it probably makes mores sense a a required argument anyway, since it is important to be intentional for sanitization.